harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -636,3 +666,77 @@ def test_checkpoint_loading(ckpt: str, istrial: bool):
         assert isinstance(trial, pytorch.PyTorchTrial), type(trial)
     else:
         assert isinstance(trial, torch.nn.Module), type(trial)
+
+
+def make_amp_workloads(
+    trial_class,
+    assert_output_float16=False,
+) -> workload.Stream:
+    trainer = utils.TrainAndValidate()
+    yield from trainer.send(steps=20, validation_freq=1, scheduling_unit=1)
+    training_metrics, _ = trainer.result()
+
+    for metrics, next_metrics in zip(training_metrics[:-1], training_metrics[1:]):
+        scale = next_metrics["scale_before"]
+        if "scale" in metrics:
```

Comment:
I think this is another case where it's safe to make stronger assumptions about the behavior of the code we're dealing with, since we are writing the tests.

Basically, just write the test it so that the metrics you want are always present in all cases, and skip the if statements and complex fallback logic.

Reply:
I think this is as strong as it can be. Is the `"scale" in metrics` statement what's bothering you?

Reply:
yes, why would "scale" not be in metrics?

Reply:
If it's unknown, which it is in the Apex case? We could write `metrics["scale"] = None` I suppose. Then of course the logic statement becomes `metrics["scale"] is None`. 

I see the advantage I think you have mind: that way, `metrics` has the same set of keys no matter what.

The disadvantage I've had in mind depends on how strict one wants to be about the (ab)use of `None`. The scale isn't nothing, it's just not known by the code that creates the `metrics` object.

I'm willing to skirt the semantics debate and go with your suggestion (assuming I understand you correctly). ;-p

Reply:
Oops, writing `metrics["scale"] = None` in `_train_batch` causes a `ZeroDivisionError` on the line `metrics = det.util.make_metrics(num_inputs, per_batch_metrics)` in `PyTorchTrialController._train_for_step`. Numpy is trying to take the mean of an empty list. I'm not sure why this causes an error instead of just giving `nan`. :-(

Reply:
> If it's unknown, which it is in the Apex case

Oh, I didn't realize that.  If it's not known in the apex case, then I'm happy with what you had.

Reply:
In defense of the offended code, bunching the scale and stage into `metrics` _is_ an abuse, since those aren't metrics per se but are rather state variables.